File: sections/69_master_structures_trees_mixed/problems/d_beacon_distance_sum/solution.py
import sys
from array import array


def main():
    t = sys.stdin.buffer.read().split()
    if not t:
        return
    z = 0
    n = int(t[z])
    q = int(t[z + 1])
    z += 2
    g = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = int(t[z]) - 1
        v = int(t[z + 1]) - 1
        z += 2
        g[u].append(v)
        g[v].append(u)
    blocked = [False] * n
    subtree = [0] * n
    paths = [array("i") for _ in range(n)]
    todo = [0]
    while todo:
        entry = todo.pop()
        nodes = [entry]
        parent = {entry: -1}
        for u in nodes:
            for v in g[u]:
                if not blocked[v] and v not in parent:
                    parent[v] = u
                    nodes.append(v)
        for u in reversed(nodes):
            subtree[u] = 1 + sum(subtree[v] for v in g[u] if parent.get(v) == u)
        total = len(nodes)
        centroid = entry
        for u in nodes:
            largest = total - subtree[u]
            for v in g[u]:
                if not blocked[v] and parent.get(v) == u:
                    largest = max(largest, subtree[v])
            if largest * 2 <= total:
                centroid = u
                break
        paths[centroid].extend((centroid, 0, -1))
        for first in g[centroid]:
            if not blocked[first]:
                stack = [(first, centroid, 1)]
                while stack:
                    u, p, d = stack.pop()
                    paths[u].extend((centroid, d, first))
                    for v in g[u]:
                        if v != p and not blocked[v]:
                            stack.append((v, u, d + 1))
        blocked[centroid] = True
        for v in g[centroid]:
            if not blocked[v]:
                todo.append(v)
    count = [0] * n
    distance_sum = [0] * n
    excluded = {}
    active = [False] * n
    out = []
    for _ in range(q):
        kind = t[z].decode()
        u = int(t[z + 1]) - 1
        z += 2
        if kind == "T":
            delta = -1 if active[u] else 1
            active[u] = not active[u]
            path = paths[u]
            for i in range(0, len(path), 3):
                c, d, b = path[i], path[i + 1], path[i + 2]
                count[c] += delta
                distance_sum[c] += delta * d
                if b != -1:
                    old_count, old_sum = excluded.get((c, b), (0, 0))
                    excluded[c, b] = (old_count + delta, old_sum + delta * d)
        else:
            answer = 0
            path = paths[u]
            for i in range(0, len(path), 3):
                c, d, b = path[i], path[i + 1], path[i + 2]
                answer += distance_sum[c] + count[c] * d
                if b != -1:
                    old_count, old_sum = excluded.get((c, b), (0, 0))
                    answer -= old_sum + old_count * d
            out.append(str(answer))
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/69_master_structures_trees_mixed/problems/d_beacon_distance_sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
struct Entry {
  int centroid, distance, branch;
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  vector<vector<int>> g(n);
  for (int i = 1, u, v; i < n; i++) {
    cin >> u >> v;
    --u;
    --v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  vector<char> blocked(n), active(n);
  vector<int> subtree(n);
  vector<int> traversal_parent(n), seen(n);
  int stamp = 0;
  vector<vector<Entry>> path(n);
  vector<int> todo{0};
  while (!todo.empty()) {
    int entry = todo.back();
    todo.pop_back();
    ++stamp;
    vector<int> nodes{entry};
    seen[entry] = stamp;
    traversal_parent[entry] = -1;
    for (int i = 0; i < (int)nodes.size(); i++) {
      int u = nodes[i];
      for (int v : g[u])
        if (!blocked[v] && seen[v] != stamp) {
          seen[v] = stamp;
          traversal_parent[v] = u;
          nodes.push_back(v);
        }
    }
    for (int i = nodes.size() - 1; i >= 0; i--) {
      int u = nodes[i];
      subtree[u] = 1;
      for (int v : g[u])
        if (!blocked[v] && seen[v] == stamp && traversal_parent[v] == u)
          subtree[u] += subtree[v];
    }
    int total = nodes.size(), centroid = entry;
    for (int u : nodes) {
      int largest = total - subtree[u];
      for (int v : g[u])
        if (!blocked[v] && seen[v] == stamp && traversal_parent[v] == u)
          largest = max(largest, subtree[v]);
      if (largest * 2 <= total) {
        centroid = u;
        break;
      }
    }
    path[centroid].push_back({centroid, 0, -1});
    for (int first : g[centroid])
      if (!blocked[first]) {
        vector<tuple<int, int, int>> stack{{first, centroid, 1}};
        while (!stack.empty()) {
          auto [u, p, d] = stack.back();
          stack.pop_back();
          path[u].push_back({centroid, d, first});
          for (int v : g[u])
            if (v != p && !blocked[v]) stack.push_back({v, u, d + 1});
        }
      }
    blocked[centroid] = 1;
    for (int v : g[centroid])
      if (!blocked[v]) todo.push_back(v);
  }
  vector<int64> count(n), sum(n);
  unordered_map<long long, pair<int64, int64>> excluded;
  auto key = [&](int c, int b) { return 1LL * c * n + b; };
  while (q--) {
    char type;
    int u;
    cin >> type >> u;
    --u;
    if (type == 'T') {
      int delta = active[u] ? -1 : 1;
      active[u] ^= 1;
      for (auto [c, d, b] : path[u]) {
        count[c] += delta;
        sum[c] += 1LL * delta * d;
        if (b != -1) {
          auto& state = excluded[key(c, b)];
          state.first += delta;
          state.second += 1LL * delta * d;
        }
      }
    } else {
      int64 answer = 0;
      for (auto [c, d, b] : path[u]) {
        answer += sum[c] + count[c] * d;
        if (b != -1) {
          auto state = excluded[key(c, b)];
          answer -= state.second + state.first * d;
        }
      }
      cout << answer << '\n';
    }
  }
}
